packages/fpdart/lib/src/effect.dart

```diff
@@ -0,0 +1,800 @@
+                  return Left<Cause<L>, R>(error.cause);
+                }
+
+                return Left<Cause<L>, R>(Die(error, stackTrace));
+              },
+            );
+          } on _EffectThrow<L> catch (genError) {
+            return Left(genError.cause);
+          }
+        },
+      );
+
+  /// {@category constructors}
+  factory Effect.tryCatch({
+    required FutureOr<R> Function() execute,
+    required L Function(Object error, StackTrace stackTrace) onError,
+    FutureOr<dynamic> Function()? onCancel,
+  }) =>
+      Effect.from(
+        (env) {
+          try {
+            return execute().then(
+              Right.new,
+              onError: (error, stackTrace) => Left(
+                Failure(onError(error, stackTrace), stackTrace),
+              ),
+            );
+          } catch (err, stack) {
+            return Left(Failure(onError(err, stack), stack));
+          }
+        },
+      );
+
+  /// {@category constructors}
+  factory Effect.fromNullable(R? value, {required L Function() onNull}) =>
+      Effect.from(
+        (_) => value == null ? Left(Failure(onNull())) : Right(value),
+      );
+
+  /// {@category constructors}
+  factory Effect.functionFail(FutureOr<Cause<L>> Function() f) => Effect.from(
+        (_) => f().then(Left.new),
+      );
+
+  /// {@category constructors}
+  factory Effect.functionSucceed(FutureOr<R> Function() f) => Effect.from(
+        (_) => f().then(Right.new),
+      );
+
+  /// {@category constructors}
+  factory Effect.lazy(Effect<E, L, R> Function() effect) =>
+      Effect.from((context) => effect().__unsafeRun(context));
+
+  /// {@category constructors}
+  factory Effect.fail(L value) => Effect.from((_) => Left(Failure(value)));
+
+  /// {@category constructors}
+  factory Effect.failCause(Cause<L> cause) => Effect.from((_) => Left(cause));
+
+  /// {@category constructors}
+  factory Effect.succeed(R value) => Effect.from((_) => Right(value));
+
+  /// {@category constructors}
+  factory Effect.raceAll(Iterable<Effect<E, L, R>> iterable) =>
+      Effect.from((context) {
+        final signal = Deferred<Never, Never>.unsafeMake();
+        final deferred = Deferred<L, R>.unsafeMake();
+
+        for (final effect in iterable) {
+          effect
+              .__unsafeRun(context.withSignal(signal))
+              .then(deferred.unsafeCompleteExit);
+
+          if (deferred.unsafeCompleted) {
+            break;
+          }
+        }
+
+        return deferred.wait<E>().__unsafeRun(context).then(
+              (exit) => signal
+                  .failCause<E, L>(const Interrupted())
+                  .__unsafeRun(context.withoutSignal)
+                  .then((_) => exit),
+            );
+      });
+
+  /// {@category constructors}
+  static Effect<E, L, void> sleep<E, L>(Duration duration) => Effect.from(
+        (_) => Future.delayed(
+          duration,
+          () => const Right(null),
+        ),
+      );
```

Comment:
It is possible to use `Timer`+`Completer`.

@tim-smart How would interruption work for `sleep`? How can `sleep` detect an interrupt and block timer?

```suggestion
  static Effect<E, L, void> sleep<E, L>(Duration duration) => Effect.from(
        (_) {
          final completer = Completer<Exit<L, void>>();
          Timer(duration, () {
            completer.complete(const Right(null));
          });

          return completer.future;
        },
      );
```

Reply:
First you will want to add a Effect.async/asyncInterrupt for callback apis, then use Effect.asyncInterrupt with the Timer api. Then hook up timer.cancel.

Reply:
@tim-smart [I added `async`](https://github.com/SandroMaglione/fpdart/pull/147/commits/1118bd66039da1ac926b178ab347be92117b5df8), but I am having some issues understanding the mental model for `Deferred` and interruption.

Could you give me an example of how to manually interrupt an `Effect` with code? Does this required `Deferred` or is it about [`Context`](https://github.com/SandroMaglione/fpdart/blob/effect-refactor/packages/fpdart/lib/src/context.dart)?

Reply:
Sleep should be:

```dart
  static Effect<E, L, void> sleep<E, L>(Duration duration) =>
      Effect.asyncInterrupt(
        (resume) {
          final timer = Timer(duration, () {
            resume.succeed(null);
          });
          return Effect.succeedLazy(() {
            timer.cancel();
          });
        },
      );
```

Reply:
As for interrupting an Effect, here is some pseudo code:

```dart
final cancel = Effect.never.runCallback();

// some time later...

cancel();
```

Reply:
@tim-smart what is `runCallback` in the example? What is the equivalent in `Effect`?

Is `cancel` a `Deferred`? Should there be any API that returns a function to interrupt the `Effect`?

Reply:
https://effect-ts.github.io/effect/effect/Effect.ts.html#runcallback

In dart maybe something like:

```dart
void Function() runCallback({ void Function(Exit<E, A> exit)? onExit })
```

Reply:
Is `runCallback` the only way to manually trigger an interruption?

How would you test interruption with `sleep` (or also `asyncInterrupt`)?

Reply:
In Effect (typescript) you can pass an abort signal to Effect.runPromise, use the cancel callback from Effect.runCallback or call interrupt on fiber returned from Effect.runFork.

dart doesn't have a built in "cancellation" primitive, so you could use a Future potentially.

Reply:
Do you think is it possible to use `Deferred` (or `Context`) in fpdart to give access to interruption?

Reply:
Not too sure what you mean, might just have to try it out!